From 65b1c86188179109746ef41c9284b27456ce4dd3 Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Tue, 22 Dec 2020 12:12:57 +0200
Subject: [PATCH] [SW-20154] make security.c platform specific

Most of the functions in security.c are platform
dependent. Move them to platform
specific file and override them in a new platform
as needed.
This makes easier addition of Gaudi2 support

Change-Id: Idf4a1040fcb78f36c777fcb2df732c1b58830e8d
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/Makefile b/Makefile
index 2a3789c..10d2d98 100644
--- a/Makefile
+++ b/Makefile
@@ -40,6 +40,7 @@
 export CFLAGS += -Os -fdata-sections -ffunction-sections
 
 ARCBOOT_COMMON_SRCS_C := $(shell ls common/*.c $(HL_DEV)/$(HL_TARGET).c)
+ARCBOOT_COMMON_SRCS_C += $(shell ls $(HL_DEV)/security.c)
 
 ARCBOOT_RAM_SRCS_C := app/arcboot_ram.c
 ARCBOOT_RAM_SRCS_S := app/crt0_ram.S
diff --git a/common/security.c b/common/security.c
index c19a41d..f5a589a 100644
--- a/common/security.c
+++ b/common/security.c
@@ -11,118 +11,6 @@
 #include "security.h"
 #include "spi_flash.h"
 
-static void security_disable_i2c_s_mstr_if(void)
-{
-	/* set AXUSER to shared */
-	regmap_writel(I2C_S_MSTR_IF_AXUSER_CORE_SHARED_MASK,
-	      mmI2C_S_MSTR_IF_AXUSER_CORE);
-	/* override AXI prot bit by master IF to secure, unprivlaged access */
-	regmap_writel(MSTR_IF_CORE_HBW_AXPROT_OVRD_OVRD_EN,
-		      mmI2C_S_MSTR_IF_CORE_HBW_AXPROT_OVRD);
-	/* enable access to PSOC_SECURITY only */
-	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
-		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
-		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
-	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
-		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-
-	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
-		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
-		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
-	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
-		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
-}
-
-static void security_disable_smi_mstr_if(void)
-{
-	regmap_writel(SMI_MSTR_IF_AXUSER_CORE_SHARED_MASK,
-	      mmSMI_MSTR_IF_AXUSER_CORE);
-	/* override AXI prot bit by master IF to secure, unprivlaged access */
-	regmap_writel(MSTR_IF_CORE_HBW_AXPROT_OVRD_OVRD_EN,
-		      mmSMI_MSTR_IF_CORE_HBW_AXPROT_OVRD);
-	/* enable access to PSOC_SECURITY only */
-	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
-		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
-		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
-	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
-		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-
-	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
-		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
-		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
-	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
-		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
-}
-
-static void security_disable_jtag_mstr_if(void)
-{
-	regmap_writel(JT_MSTR_IF_AXUSER_CORE_SHARED_MASK,
-		      mmJT_MSTR_IF_AXUSER_CORE);
-
-	/* override AXI prot bit by master IF to secure, unprivlaged access */
-	regmap_writel(MSTR_IF_CORE_HBW_AXPROT_OVRD_OVRD_EN,
-		      mmJT_MSTR_IF_CORE_HBW_AXPROT_OVRD);
-	/* enable access to PSOC_SECURITY only */
-	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
-	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-
-	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
-	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
-}
-
-static void security_enable_jtag_mstr_if(void)
-{
-	regmap_writel(0, mmJT_MSTR_IF_AXUSER_CORE);
-	/* don't override AXI prot bit by master IF */
-	regmap_writel(0, mmJT_MSTR_IF_CORE_HBW_AXPROT_OVRD);
-
-	/* enable access to all */
-	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
-	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
-		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-}
-
-void security_disable_twi_if(void)
-{
-	security_disable_i2c_s_mstr_if();
-	security_disable_smi_mstr_if();
-}
-
-static void security_jtag_dap(uint8_t status)
-{
-	/* close dap WA jira: H5-1803 */
-	regmap_writel(status, mmPSOC_CS_TRACE_SPECIAL_GLBL_SPARE_0);
-}
-
 void security_disable_jtag_if(void)
 {
 	regmap_writel(0, mmPSOC_CS_TRACE_TOP_ROM_VALID_0);
@@ -148,19 +36,6 @@
 		      mmSMI_MSTR_IF_CORE_HBW_AXPROT_OVRD);
 	regmap_writel(MSTR_IF_CORE_HBW_AXPROT_OVRD_OVRD_EN,
 		      mmJT_MSTR_IF_CORE_HBW_AXPROT_OVRD);
-}
-
-void security_debug_if_shared(void)
-{
-	/* Configure SMI,I2C and JTAG to be "shared" engine */
-	regmap_writel(I2C_S_MSTR_IF_AXUSER_CORE_SHARED_MASK,
-		      mmI2C_S_MSTR_IF_AXUSER_CORE);
-	regmap_writel(SMI_MSTR_IF_AXUSER_CORE_SHARED_MASK,
-		      mmSMI_MSTR_IF_AXUSER_CORE);
-	regmap_writel(JT_MSTR_IF_AXUSER_CORE_SHARED_MASK,
-		      mmJT_MSTR_IF_AXUSER_CORE);
-	regmap_writel(MSTR_IF_AXUSER_CORE_SHARED,
-		      mmPSOC_MSTR_IF_AXUSER_CORE);
 }
 
 uint32_t security_key_soc_location_offset(uint32_t hash_key_index)
@@ -200,7 +75,7 @@
 		      mmPSOC_GLOBAL_CONF_MEM_CPY_AXUSER);
 }
 
-static void security_elbi_rr_lbw_config(uint8_t spi_controller_en)
+void security_elbi_rr_lbw_config(uint8_t spi_controller_en)
 {
 	regmap_writel(SECURITY_PSOC_START_LBW,
 		mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_0);
@@ -224,234 +99,6 @@
 		regmap_writel(SECURITY_PSOC_END_LBW,
 			mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_1);
 	}
-}
-
-static void security_dcon_rr_lbw_config(uint8_t spi_controller_en)
-{
-	regmap_writel(SECURITY_PSOC_START_LBW,
-		mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_0);
-
-	if (spi_controller_en) {
-		regmap_writel(SECURITY_QSPI_START_LBW,
-		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_0);
-		regmap_writel(SECURITY_QSPI_END_LBW,
-		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_1);
-		regmap_writel(SECURITY_PSOC_SECURITY_START_LBW,
-		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_1);
-		regmap_writel(SECURITY_PSOC_SECURITY_END_LBW,
-		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_2);
-		regmap_writel(SECURITY_PSOC_END_LBW,
-		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_2);
-	} else {
-		regmap_writel(SECURITY_PSOC_SECURITY_START_LBW,
-		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_0);
-		regmap_writel(SECURITY_PSOC_SECURITY_END_LBW,
-		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_1);
-		regmap_writel(SECURITY_PSOC_END_LBW,
-		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_1);
-	}
-}
-
-static void security_dcon_if0_rr_hbw_config(uint8_t spi_controller_en)
-{
-	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
-		mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
-		mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
-
-	if (spi_controller_en) {
-		regmap_writel(SECURITY_QSPI_START_HBW,
-		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-		regmap_writel(SECURITY_QSPI_END_HBW,
-		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	} else {
-		regmap_writel(SECURITY_FLASH_START_HBW,
-		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-		regmap_writel(SECURITY_FLASH_END_HBW,
-		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	}
-
-	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
-		mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
-	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
-		mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
-}
-
-static void security_dcon_if1_rr_hbw_config(uint8_t spi_controller_en)
-{
-	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
-		mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
-		mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
-
-	if (spi_controller_en) {
-		regmap_writel(SECURITY_QSPI_START_HBW,
-		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-		regmap_writel(SECURITY_QSPI_END_HBW,
-		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	} else {
-		regmap_writel(SECURITY_FLASH_START_HBW,
-		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-		regmap_writel(SECURITY_FLASH_END_HBW,
-		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	}
-
-	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
-		mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
-	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
-		mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
-}
-
-static void security_i2c_s_rr_hbw_config(uint8_t spi_controller_en)
-{
-	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
-		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
-		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
-
-	if (spi_controller_en) {
-		regmap_writel(SECURITY_QSPI_START_HBW,
-			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-		regmap_writel(SECURITY_QSPI_END_HBW,
-			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	} else {
-		regmap_writel(SECURITY_FLASH_START_HBW,
-			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-		regmap_writel(SECURITY_FLASH_END_HBW,
-			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	}
-
-	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
-		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
-	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
-		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_2);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_2);
-	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
-		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_2);
-	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
-		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_2);
-}
-
-static void security_smi_rr_hbw_config(uint8_t spi_controller_en)
-{
-	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
-		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
-		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
-
-	if (spi_controller_en) {
-		regmap_writel(SECURITY_QSPI_START_HBW,
-			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-		regmap_writel(SECURITY_QSPI_END_HBW,
-			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	} else {
-		regmap_writel(SECURITY_FLASH_START_HBW,
-			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-		regmap_writel(SECURITY_FLASH_END_HBW,
-			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	}
-
-	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
-		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
-	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
-		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_2);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_2);
-	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
-		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_2);
-	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
-		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_2);
-}
-
-static void security_jtag_rr_hbw_config(uint8_t spi_controller_en)
-{
-	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
-		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
-		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
-
-	if (spi_controller_en) {
-		regmap_writel(SECURITY_QSPI_START_HBW,
-			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-		regmap_writel(SECURITY_QSPI_END_HBW,
-			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	} else {
-		regmap_writel(SECURITY_FLASH_START_HBW,
-			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
-		regmap_writel(SECURITY_FLASH_END_HBW,
-			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
-		regmap_writel(SECURITY_HBW_MSB_ZERO,
-			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
-	}
-
-	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
-		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
-	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
-		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_2);
-	regmap_writel(SECURITY_HBW_MSB_ZERO,
-		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_2);
-	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
-		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_2);
-	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
-		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_2);
-}
-
-void security_pcie_rr_config(uint8_t spi_en)
-{
-	regmap_writel(PCIE_WRAP_PCIE_PROT_OVR_WR_EN_MASK |
-		      PCIE_WRAP_PCIE_PROT_OVR_RD_EN_MASK,
-		      mmPCIE_WRAP_PCIE_PROT_OVR);
-	regmap_writel(PCIE_MSTR_RR_MSTR_IF_AXUSER_CORE_SHARED_MASK,
-		      mmPCIE_MSTR_RR_MSTR_IF_AXUSER_CORE);
-	security_elbi_rr_lbw_config(spi_en);
-	security_dcon_rr_lbw_config(spi_en);
-	security_dcon_if0_rr_hbw_config(spi_en);
-	security_dcon_if1_rr_hbw_config(spi_en);
 }
 
 /* If DBG I/F were not disabled, program the RR to allow access
diff --git a/common/security.h b/common/security.h
index 90a8991..819adec 100644
--- a/common/security.h
+++ b/common/security.h
@@ -128,4 +128,12 @@
 				enum security_img_status img_stat);
 void security_report_status(uint8_t index, enum security_status sec_stat);
 
+void security_disable_jtag_mstr_if(void);
+void security_enable_jtag_mstr_if(void);
+void security_i2c_s_rr_hbw_config(uint8_t spi_controller_en);
+void security_smi_rr_hbw_config(uint8_t spi_controller_en);
+void security_jtag_rr_hbw_config(uint8_t spi_controller_en);
+void security_jtag_dap(uint8_t status);
+void security_elbi_rr_lbw_config(uint8_t spi_controller_en);
+
 #endif /* _SECURITY_H_ */
diff --git a/goya2_arc/security.c b/goya2_arc/security.c
new file mode 100644
index 0000000..d9c290d
--- /dev/null
+++ b/goya2_arc/security.c
@@ -0,0 +1,365 @@
+/*
+ * Copyright 2016-2019 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
+ */
+
+#include "arcboot.h"
+#include "security.h"
+#include "efuse.h"
+#include "security.h"
+#include "spi_flash.h"
+
+static void security_disable_i2c_s_mstr_if(void)
+{
+	/* set AXUSER to shared */
+	regmap_writel(I2C_S_MSTR_IF_AXUSER_CORE_SHARED_MASK,
+	      mmI2C_S_MSTR_IF_AXUSER_CORE);
+	/* override AXI prot bit by master IF to secure, unprivlaged access */
+	regmap_writel(MSTR_IF_CORE_HBW_AXPROT_OVRD_OVRD_EN,
+		      mmI2C_S_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+	/* enable access to PSOC_SECURITY only */
+	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
+		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
+	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
+		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
+	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
+		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+
+	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
+		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
+		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
+	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
+		      mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
+}
+
+static void security_disable_smi_mstr_if(void)
+{
+	regmap_writel(SMI_MSTR_IF_AXUSER_CORE_SHARED_MASK,
+	      mmSMI_MSTR_IF_AXUSER_CORE);
+	/* override AXI prot bit by master IF to secure, unprivlaged access */
+	regmap_writel(MSTR_IF_CORE_HBW_AXPROT_OVRD_OVRD_EN,
+		      mmSMI_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+	/* enable access to PSOC_SECURITY only */
+	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
+		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
+	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
+		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
+	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
+		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+
+	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
+		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
+		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
+	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
+		      mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
+}
+
+void security_disable_jtag_mstr_if(void)
+{
+	regmap_writel(JT_MSTR_IF_AXUSER_CORE_SHARED_MASK,
+		      mmJT_MSTR_IF_AXUSER_CORE);
+
+	/* override AXI prot bit by master IF to secure, unprivlaged access */
+	regmap_writel(MSTR_IF_CORE_HBW_AXPROT_OVRD_OVRD_EN,
+		      mmJT_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+	/* enable access to PSOC_SECURITY only */
+	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
+	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
+	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+
+	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
+	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
+}
+
+void security_enable_jtag_mstr_if(void)
+{
+	regmap_writel(0, mmJT_MSTR_IF_AXUSER_CORE);
+	/* don't override AXI prot bit by master IF */
+	regmap_writel(0, mmJT_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+
+	/* enable access to all */
+	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
+	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
+	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
+		      mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+}
+
+void security_disable_twi_if(void)
+{
+	security_disable_i2c_s_mstr_if();
+	security_disable_smi_mstr_if();
+}
+
+void security_jtag_dap(uint8_t status)
+{
+	/* close dap WA jira: H5-1803 */
+	regmap_writel(status, mmPSOC_CS_TRACE_SPECIAL_GLBL_SPARE_0);
+}
+
+void security_debug_if_shared(void)
+{
+	/* Configure SMI,I2C and JTAG to be "shared" engine */
+	regmap_writel(I2C_S_MSTR_IF_AXUSER_CORE_SHARED_MASK,
+		      mmI2C_S_MSTR_IF_AXUSER_CORE);
+	regmap_writel(SMI_MSTR_IF_AXUSER_CORE_SHARED_MASK,
+		      mmSMI_MSTR_IF_AXUSER_CORE);
+	regmap_writel(JT_MSTR_IF_AXUSER_CORE_SHARED_MASK,
+		      mmJT_MSTR_IF_AXUSER_CORE);
+	regmap_writel(MSTR_IF_AXUSER_CORE_SHARED,
+		      mmPSOC_MSTR_IF_AXUSER_CORE);
+}
+
+static void security_dcon_rr_lbw_config(uint8_t spi_controller_en)
+{
+	regmap_writel(SECURITY_PSOC_START_LBW,
+		mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_0);
+
+	if (spi_controller_en) {
+		regmap_writel(SECURITY_QSPI_START_LBW,
+		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_0);
+		regmap_writel(SECURITY_QSPI_END_LBW,
+		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_1);
+		regmap_writel(SECURITY_PSOC_SECURITY_START_LBW,
+		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_1);
+		regmap_writel(SECURITY_PSOC_SECURITY_END_LBW,
+		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_2);
+		regmap_writel(SECURITY_PSOC_END_LBW,
+		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_2);
+	} else {
+		regmap_writel(SECURITY_PSOC_SECURITY_START_LBW,
+		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_0);
+		regmap_writel(SECURITY_PSOC_SECURITY_END_LBW,
+		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_1);
+		regmap_writel(SECURITY_PSOC_END_LBW,
+		       mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_1);
+	}
+}
+
+static void security_dcon_if0_rr_hbw_config(uint8_t spi_controller_en)
+{
+	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
+		mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
+	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
+		mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
+
+	if (spi_controller_en) {
+		regmap_writel(SECURITY_QSPI_START_HBW,
+		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+		regmap_writel(SECURITY_QSPI_END_HBW,
+		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	} else {
+		regmap_writel(SECURITY_FLASH_START_HBW,
+		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+		regmap_writel(SECURITY_FLASH_END_HBW,
+		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+		   mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	}
+
+	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
+		mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
+	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
+		mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
+}
+
+static void security_dcon_if1_rr_hbw_config(uint8_t spi_controller_en)
+{
+	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
+		mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
+	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
+		mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
+
+	if (spi_controller_en) {
+		regmap_writel(SECURITY_QSPI_START_HBW,
+		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+		regmap_writel(SECURITY_QSPI_END_HBW,
+		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	} else {
+		regmap_writel(SECURITY_FLASH_START_HBW,
+		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+		regmap_writel(SECURITY_FLASH_END_HBW,
+		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+		   mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	}
+
+	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
+		mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
+	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
+		mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
+}
+
+void security_i2c_s_rr_hbw_config(uint8_t spi_controller_en)
+{
+	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
+		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
+	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
+		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
+
+	if (spi_controller_en) {
+		regmap_writel(SECURITY_QSPI_START_HBW,
+			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+		regmap_writel(SECURITY_QSPI_END_HBW,
+			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	} else {
+		regmap_writel(SECURITY_FLASH_START_HBW,
+			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+		regmap_writel(SECURITY_FLASH_END_HBW,
+			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	}
+
+	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
+		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
+	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
+		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_2);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_2);
+	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
+		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_2);
+	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
+		mmI2C_S_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_2);
+}
+
+void security_smi_rr_hbw_config(uint8_t spi_controller_en)
+{
+	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
+		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
+	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
+		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
+
+	if (spi_controller_en) {
+		regmap_writel(SECURITY_QSPI_START_HBW,
+			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+		regmap_writel(SECURITY_QSPI_END_HBW,
+			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	} else {
+		regmap_writel(SECURITY_FLASH_START_HBW,
+			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+		regmap_writel(SECURITY_FLASH_END_HBW,
+			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	}
+
+	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
+		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
+	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
+		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_2);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_2);
+	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
+		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_2);
+	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
+		mmSMI_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_2);
+}
+
+void security_jtag_rr_hbw_config(uint8_t spi_controller_en)
+{
+	regmap_writel(SECURITY_PSOC_START_HBW_LSB,
+		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0);
+	regmap_writel(SECURITY_PSOC_START_HBW_MSB,
+		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0);
+
+	if (spi_controller_en) {
+		regmap_writel(SECURITY_QSPI_START_HBW,
+			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+		regmap_writel(SECURITY_QSPI_END_HBW,
+			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	} else {
+		regmap_writel(SECURITY_FLASH_START_HBW,
+			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0);
+		regmap_writel(SECURITY_FLASH_END_HBW,
+			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_1);
+		regmap_writel(SECURITY_HBW_MSB_ZERO,
+			mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_1);
+	}
+
+	regmap_writel(SECURITY_PSOC_SECURITY_START_HBW,
+		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_1);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_1);
+	regmap_writel(SECURITY_PSOC_SECURITY_END_HBW,
+		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_2);
+	regmap_writel(SECURITY_HBW_MSB_ZERO,
+		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_2);
+	regmap_writel(SECURITY_PSOC_END_HBW_LSB,
+		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_2);
+	regmap_writel(SECURITY_PSOC_END_HBW_MSB,
+		mmJT_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_2);
+}
+
+void security_pcie_rr_config(uint8_t spi_en)
+{
+	regmap_writel(PCIE_WRAP_PCIE_PROT_OVR_WR_EN_MASK |
+		      PCIE_WRAP_PCIE_PROT_OVR_RD_EN_MASK,
+		      mmPCIE_WRAP_PCIE_PROT_OVR);
+	regmap_writel(PCIE_MSTR_RR_MSTR_IF_AXUSER_CORE_SHARED_MASK,
+		      mmPCIE_MSTR_RR_MSTR_IF_AXUSER_CORE);
+	security_elbi_rr_lbw_config(spi_en);
+	security_dcon_rr_lbw_config(spi_en);
+	security_dcon_if0_rr_hbw_config(spi_en);
+	security_dcon_if1_rr_hbw_config(spi_en);
+}
